README.md
# Rust VEML6040 RGBW Color Sensor Driver

[![crates.io](https://img.shields.io/crates/v/veml6040.svg)](https://crates.io/crates/veml6040)
[![Docs](https://docs.rs/veml6040/badge.svg)](https://docs.rs/veml6040)
[![Build Status](https://github.com/eldruin/veml6040-rs/workflows/Build/badge.svg)](https://github.com/eldruin/veml6040-rs/actions?query=workflow%3ABuild)
[![Coverage Status](https://coveralls.io/repos/github/eldruin/veml6040-rs/badge.svg?branch=master)](https://coveralls.io/github/eldruin/veml6040-rs?branch=master)

This is a platform agnostic Rust driver for the VEML6040 RGBW color light
sensor, based on the [`embedded-hal`] traits.

[`embedded-hal`]: https://github.com/rust-embedded/embedded-hal

This driver allows you to:
- Enable/disable the sensor.
- Set the integration time.
- Set the measurement mode.
- Trigger a measurement when on manual mode.
- Read the red channel measurement.
- Read the green channel measurement.
- Read the blue channel measurement.
- Read the white channel measurement.
- Read measurement of all channels at once.

## The device

VEML6040 color sensor senses red, green, blue, and white light and
incorporates photodiodes, amplifiers, and analog / digital circuits into a
single chip using CMOS process. With the color sensor applied, the
brightness, and color temperature of backlight can be adjusted base on
ambient light source that makes panel looks more comfortable for end
user's eyes. VEML6040's adoption of Filtron TM technology achieves the
closest ambient light spectral sensitivity to real human eye responses.
VEML6040 provides excellent temperature compensation capability for keeping
the output stable under changing temperature. VEML6040's function are
easily operated via the simple command format of I2C (SMBus compatible)
interface protocol. VEML6040's operating voltage ranges from 2.5 V to 3.6 V.

Datasheet: [VEML6040](https://www.vishay.com/docs/84276/veml6040.pdf)

Application note: [VEML6040 AN](https://www.vishay.com/docs/84331/designingveml6040.pdf)

## Usage

To use this driver, import this crate and an `embedded_hal` implementation,
then instantiate the device.

Please find additional examples using hardware in this repository: [driver-examples]

[driver-examples]: https://github.com/eldruin/driver-examples

```rust
use linux_embedded_hal::I2cdev;
use veml6040::Veml6040;

fn main() {
    let dev = I2cdev::new("/dev/i2c-1").unwrap();
    let mut sensor = Veml6040::new(dev);
    sensor.enable().unwrap();

    let red = sensor.read_red_channel().unwrap();
    let green = sensor.read_green_channel().unwrap();
    let blue = sensor.read_blue_channel().unwrap();
    let white = sensor.read_white_channel().unwrap();

    println!(
        "Measurements: R: {}, G: {}, B: {}, W: {}",
        red, green, blue, white
    );
}
```

## Support

For questions, issues, feature requests, and other changes, please file an
[issue in the github project](https://github.com/eldruin/veml6040-rs/issues).

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT)
   
at your option.

### Contributing

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

...

## [0.1.1] - 2021-01-25
### Changed
- Make types `IntegrationTime` and `MeasurementMode` copyable.
- Make type `AllChannelMeasurement` debuggable, clonable, copyable and comparable.

## 0.1.0 - 2018-10-23

This is the initial release to crates.io of the feature-complete driver. There
may be some API changes in the future, in case I decide that something can be
further improved. All changes will be documented in this CHANGELOG.

[Unreleased]: https://github.com/eldruin/veml6040-rs/compare/v0.1.1...HEAD
[0.1.1]: https://github.com/eldruin/embedded-ccs811-rs/compare/v0.1.0...v0.1.1



tests/reading.rs
use embedded_hal_mock::i2c::Transaction as I2cTransaction;
mod common;
use self::common::{setup, Register, DEVICE_ADDRESS};

fn get_expectation(register: u8) -> [I2cTransaction; 1] {
    [I2cTransaction::write_read(
        DEVICE_ADDRESS,
        vec![register],
        vec![0xCD, 0xAB],
    )]
}

macro_rules! read_test {
    ($name:ident, $method:ident, $register:ident) => {
        #[test]
        fn $name() {
            let expectations = get_expectation(Register::$register);
            let mut dev = setup(&expectations);
            let data = dev.$method().unwrap();
            assert_eq!(0xABCD, data);
            dev.destroy().done();
        }
    };
}

read_test!(can_read_red_channel, read_red_channel, R_DATA);
read_test!(can_read_green_channel, read_green_channel, G_DATA);
read_test!(can_read_blue_channel, read_blue_channel, B_DATA);
read_test!(can_read_white_channel, read_white_channel, W_DATA);

#[test]
fn can_read_all_channels() {
    let expectations = [I2cTransaction::write_read(
        DEVICE_ADDRESS,
        vec![Register::R_DATA],
        vec![0x23, 0x01, 0x67, 0x45, 0xAB, 0x89, 0xEF, 0xCD],
    )];
    let mut dev = setup(&expectations);
    let measurement = dev.read_all_channels().unwrap();
    assert_eq!(0x0123, measurement.red);
    assert_eq!(0x4567, measurement.green);
    assert_eq!(0x89AB, measurement.blue);
    assert_eq!(0xCDEF, measurement.white);
    dev.destroy().done();
}


tests/common/mod.rs
use embedded_hal_mock::i2c::{Mock as I2cMock, Transaction as I2cTransaction};
use veml6040::Veml6040;

pub const DEVICE_ADDRESS: u8 = 0x10;

pub struct Register;

#[allow(unused)]
impl Register {
    pub const CONFIG: u8 = 0x00;
    pub const R_DATA: u8 = 0x08;
    pub const G_DATA: u8 = 0x09;
    pub const B_DATA: u8 = 0x0A;
    pub const W_DATA: u8 = 0x0B;
}

pub fn setup(expectations: &[I2cTransaction]) -> Veml6040<I2cMock> {
    let i2c = I2cMock::new(expectations);
    Veml6040::new(i2c)
}


tests/configuration.rs
use embedded_hal_mock::i2c::Transaction as I2cTransaction;
use veml6040::{IntegrationTime, MeasurementMode};
mod common;
use self::common::{setup, Register, DEVICE_ADDRESS};

fn get_write_expectation(config: u8) -> [I2cTransaction; 1] {
    [I2cTransaction::write(
        DEVICE_ADDRESS,
        vec![Register::CONFIG, config, 0],
    )]
}

macro_rules! config_test {
    ($name:ident, $method:ident, $expected:expr) => {
        #[test]
        fn $name() {
            let expectations = get_write_expectation($expected);
            let mut dev = setup(&expectations);
            dev.$method().unwrap();
            dev.destroy().done();
        }
    };
}

config_test!(can_enable, enable, 0);
config_test!(can_disable, disable, 1);

macro_rules! config_param_test {
    ($name:ident, $method:ident, $input:expr, $expected:expr) => {
        #[test]
        fn $name() {
            let expectations = get_write_expectation($expected);
            let mut dev = setup(&expectations);
            dev.$method($input).unwrap();
            dev.destroy().done();
        }
    };
}

config_param_test!(set_mm_auto, set_measurement_mode, MeasurementMode::Auto, 0);
config_param_test!(set_mm_man, set_measurement_mode, MeasurementMode::Manual, 2);

config_test!(can_trigger_measurement, trigger_measurement, 4);

macro_rules! set_it_test {
    ($name:ident, $variant:ident, $expected:expr) => {
        config_param_test!(
            $name,
            set_integration_time,
            IntegrationTime::$variant,
            $expected
        );
    };
}

set_it_test!(can_set_it_40, _40ms, 0);
set_it_test!(can_set_it_80, _80ms, 1 << 4);
set_it_test!(can_set_it_160, _160ms, 2 << 4);
set_it_test!(can_set_it_320, _320ms, 3 << 4);
set_it_test!(can_set_it_640, _640ms, 4 << 4);
set_it_test!(can_set_it_1280, _1280ms, 5 << 4);


src/reading.rs
use crate::{AllChannelMeasurement, Error, Register, Veml6040, DEVICE_ADDRESS};
use embedded_hal::blocking::i2c;

impl<I2C, E> Veml6040<I2C>
where
    I2C: i2c::WriteRead<Error = E>,
{
    /// Read the red channel measurement data.
    pub fn read_red_channel(&mut self) -> Result<u16, Error<E>> {
        self.read_channel(Register::R_DATA)
    }

    /// Read the green channel measurement data.
    pub fn read_green_channel(&mut self) -> Result<u16, Error<E>> {
        self.read_channel(Register::G_DATA)
    }

    /// Read the blue channel measurement data.
    pub fn read_blue_channel(&mut self) -> Result<u16, Error<E>> {
        self.read_channel(Register::B_DATA)
    }

    /// Read the white channel measurement data.
    pub fn read_white_channel(&mut self) -> Result<u16, Error<E>> {
        self.read_channel(Register::W_DATA)
    }

    /// Read the measurement data of all channels at once.
    pub fn read_all_channels(&mut self) -> Result<AllChannelMeasurement, Error<E>> {
        let mut data = [0; 8];
        self.i2c
            .write_read(DEVICE_ADDRESS, &[Register::R_DATA], &mut data)
            .map_err(Error::I2C)?;
        Ok(AllChannelMeasurement {
            red: u16::from(data[1]) << 8 | u16::from(data[0]),
            green: u16::from(data[3]) << 8 | u16::from(data[2]),
            blue: u16::from(data[5]) << 8 | u16::from(data[4]),
            white: u16::from(data[7]) << 8 | u16::from(data[6]),
        })
    }

    fn read_channel(&mut self, first_register: u8) -> Result<u16, Error<E>> {
        let mut data = [0; 2];
        self.i2c
            .write_read(DEVICE_ADDRESS, &[first_register], &mut data)
            .map_err(Error::I2C)
            .and(Ok(u16::from(data[1]) << 8 | u16::from(data[0])))
    }
}


src/configuration.rs
use crate::{
    BitFlags, Error, IntegrationTime, MeasurementMode, Register, Veml6040, DEVICE_ADDRESS,
};
use embedded_hal::blocking::i2c;

impl<I2C, E> Veml6040<I2C>
where
    I2C: i2c::Write<Error = E>,
{
    /// Enable the sensor.
    pub fn enable(&mut self) -> Result<(), Error<E>> {
        let config = self.config;
        self.write_config(config & !BitFlags::SHUTDOWN)
    }

    /// Disable the sensor (shutdown).
    pub fn disable(&mut self) -> Result<(), Error<E>> {
        let config = self.config;
        self.write_config(config | BitFlags::SHUTDOWN)
    }

    /// Set the integration time.
    pub fn set_integration_time(&mut self, it: IntegrationTime) -> Result<(), Error<E>> {
        const IT_BITS: u8 = 0b0111_0000;
        let config = self.config & !IT_BITS;
        match it {
            IntegrationTime::_40ms => self.write_config(config),
            IntegrationTime::_80ms => self.write_config(config | 0b0001_0000),
            IntegrationTime::_160ms => self.write_config(config | 0b0010_0000),
            IntegrationTime::_320ms => self.write_config(config | 0b0011_0000),
            IntegrationTime::_640ms => self.write_config(config | 0b0100_0000),
            IntegrationTime::_1280ms => self.write_config(config | 0b0101_0000),
        }
    }

    /// Set the measurement mode: `Auto`/`Manual`.
    pub fn set_measurement_mode(&mut self, mode: MeasurementMode) -> Result<(), Error<E>> {
        let config = self.config;
        match mode {
            MeasurementMode::Auto => self.write_config(config & !BitFlags::AF),
            MeasurementMode::Manual => self.write_config(config | BitFlags::AF),
        }
    }

    /// Trigger a measurement when on `Manual` measurement mode.
    ///
    /// This is not necessary on `Auto` measurement mode.
    pub fn trigger_measurement(&mut self) -> Result<(), Error<E>> {
        // This bit is not stored to avoid unintended triggers.
        self.i2c
            .write(
                DEVICE_ADDRESS,
                &[Register::CONFIG, self.config | BitFlags::TRIG, 0],
            )
            .map_err(Error::I2C)
    }

    fn write_config(&mut self, config: u8) -> Result<(), Error<E>> {
        self.i2c
            .write(DEVICE_ADDRESS, &[Register::CONFIG, config, 0])
            .map_err(Error::I2C)?;
        self.config = config;
        Ok(())
    }
}


src/lib.rs
//! This is a platform agnostic Rust driver for the VEML6040 RGBW color light
//! sensor, based on the [`embedded-hal`] traits.
//!
//! [`embedded-hal`]: https://github.com/rust-embedded/embedded-hal
//!
//! This driver allows you to:
//! - Enable/disable the sensor.
//! - Set the integration time.
//! - Set the measurement mode.
//! - Trigger a measurement when on manual mode.
//! - Read the red channel measurement.
//! - Read the green channel measurement.
//! - Read the blue channel measurement.
//! - Read the white channel measurement.
//! - Read measurement of all channels at once.
//!
//! ## The device
//!
//! VEML6040 color sensor senses red, green, blue, and white light and
//! incorporates photodiodes, amplifiers, and analog / digital circuits into a
//! single chip using CMOS process. With the color sensor applied, the
//! brightness, and color temperature of backlight can be adjusted base on
//! ambient light source that makes panel looks more comfortable for end
//! user's eyes. VEML6040's adoption of Filtron TM technology achieves the
//! closest ambient light spectral sensitivity to real human eye responses.
//! VEML6040 provides excellent temperature compensation capability for keeping
//! the output stable under changing temperature. VEML6040's function are
//! easily operated via the simple command format of I2C (SMBus compatible)
//! interface protocol. VEML6040's operating voltage ranges from 2.5 V to 3.6 V.
//!
//! Datasheet: [VEML6040](https://www.vishay.com/docs/84276/veml6040.pdf)
//!
//! Application note: [VEML6040 AN](https://www.vishay.com/docs/84331/designingveml6040.pdf)
//!
//! ## Usage examples (see also examples folder)
//!
//! To use this driver, import this crate and an `embedded_hal` implementation,
//! then instantiate the device.
//!
//! Please find additional examples using hardware in this repository: [driver-examples]
//!
//! [driver-examples]: https://github.com/eldruin/driver-examples
//!
//! ### Enable and read the color measurement
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6040::Veml6040;
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Veml6040::new(dev);
//! sensor.enable().unwrap();
//!
//! let red = sensor.read_red_channel().unwrap();
//! let green = sensor.read_green_channel().unwrap();
//! let blue = sensor.read_blue_channel().unwrap();
//! let white = sensor.read_white_channel().unwrap();
//!
//! println!("Measurements: red = {}, green = {}, blue = {}, white = {}",
//!          red, green, blue, white);
//! ```
//!
//! ### Read all the channels at once
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6040::Veml6040;
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Veml6040::new(dev);
//! sensor.enable().unwrap();
//!
//! let measurement = sensor.read_all_channels().unwrap();
//!
//! println!("Measurements: red = {}, green = {}, blue = {}, white = {}",
//!          measurement.red, measurement.green, measurement.blue,
//!          measurement.white);
//! ```
//!
//! ### Set the integration time to 320ms
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6040::{ Veml6040, IntegrationTime };
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Veml6040::new(dev);
//! sensor.enable().unwrap();
//! sensor.set_integration_time(IntegrationTime::_320ms).unwrap();
//! ```
//!
//! ### Set the measurement mode to manual and trigger a measurement
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6040::{ Veml6040, MeasurementMode };
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Veml6040::new(dev);
//! sensor.enable().unwrap();
//! sensor.set_measurement_mode(MeasurementMode::Manual).unwrap();
//! sensor.trigger_measurement().unwrap();
//! ```

#![deny(unsafe_code)]
#![deny(missing_docs)]
#![doc(html_root_url = "https://docs.rs/veml6040/0.1.1")]
#![no_std]

use embedded_hal::blocking::i2c;

/// All possible errors in this crate
#[derive(Debug)]
pub enum Error<E> {
    /// I²C bus error
    I2C(E),
}

/// Possible measurement modes
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum MeasurementMode {
    /// Automatic mode.
    ///
    /// Measurements are made continuously. The actual cadence depends on
    /// the integration time.
    Auto,
    /// Manual mode.
    ///
    /// Measurements are only triggered manually. See `trigger_measurement()`.
    /// This is also called "force mode" or "ActiveForce" mode.
    Manual,
}

/// Integration time
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum IntegrationTime {
    /// 40 ms
    _40ms,
    /// 80 ms
    _80ms,
    /// 160 ms
    _160ms,
    /// 320 ms
    _320ms,
    /// 640 ms
    _640ms,
    /// 1280 ms
    _1280ms,
}

/// Result of measurement of all channels
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct AllChannelMeasurement {
    /// Red channel measurement.
    pub red: u16,
    /// Green channel measurement.
    pub green: u16,
    /// Blue channel measurement.
    pub blue: u16,
    /// White channel measurement.
    pub white: u16,
}

const DEVICE_ADDRESS: u8 = 0x10;

struct Register;

impl Register {
    const CONFIG: u8 = 0x00;
    const R_DATA: u8 = 0x08;
    const G_DATA: u8 = 0x09;
    const B_DATA: u8 = 0x0A;
    const W_DATA: u8 = 0x0B;
}

struct BitFlags;

impl BitFlags {
    const SHUTDOWN: u8 = 0b0000_0001;
    const AF: u8 = 0b0000_0010;
    const TRIG: u8 = 0b0000_0100;
}

/// VEML6040 device driver.
#[derive(Debug, Default)]
pub struct Veml6040<I2C> {
    /// The concrete I²C device implementation.
    i2c: I2C,
    /// Configuration register status.
    config: u8,
}

impl<I2C, E> Veml6040<I2C>
where
    I2C: i2c::Write<Error = E>,
{
    /// Create new instance of the VEML6040 device.
    pub fn new(i2c: I2C) -> Self {
        Veml6040 { i2c, config: 0 }
    }

    /// Destroy driver instance, return I²C bus instance.
    pub fn destroy(self) -> I2C {
        self.i2c
    }
}

mod configuration;
mod reading;
